File: 2018/julia/day04.jl
@enum State STARTED AWAKE ASLEEP

mutable struct Event
    guard::Int
    state::State
    beg_sleep::Int
    end_sleep::Int
    Event(guard, state, beg_sleep) = new(guard, state, beg_sleep, 0)
end

vals = sort([i for i in readlines("data/input-04.txt")])

events, guards, current_guard = [], Set{Int}(), 0
for (idx, line) in enumerate(vals)
    minute = parse(Int, line[16:17])
    if idx > 1 && events[idx - 1].state == ASLEEP
        events[idx - 1].end_sleep = minute
    end
    case = line[20:24]
    if case == "falls"
        state = ASLEEP
        beg_sleep = minute
    elseif case == "wakes"
        state = AWAKE
        beg_sleep = 0
    elseif case == "Guard"
        state = STARTED
        global current_guard = parse(Int, match(r"#([0-9]+)", line).captures[1])
        push!(guards, current_guard)
        beg_sleep = 0
    else
        @assert(false, "unreachable")
    end
    push!(events, Event(current_guard, state, beg_sleep))
end

guards, sleep, minutes = collect(guards), zeros(Int, length(guards)), []
for _ in guards
    push!(minutes, zeros(Int, 60))
end
for event in events
    if event.state != ASLEEP
        continue
    end
    idx = [idx for (idx, guard) in enumerate(guards) if guard == event.guard][1]
    sleep[idx] += event.end_sleep - event.beg_sleep
    for mm in event.beg_sleep:event.end_sleep
        minutes[idx][mm + 1] += 1
    end
end

max_minute = zeros(Int, length(guards))
max_times  = zeros(Int, length(guards))
for (idx, minute) in enumerate(minutes)
    max_times[idx], max_minute[idx] = findmax(minute)
end

guard_id = argmax(sleep)
res = guards[guard_id] * (max_minute[guard_id] - 1)
println("Part 1: $(res)")
@assert(res == 12169)

guard_id = argmax(max_times)
res = guards[guard_id] * (max_minute[guard_id] - 1)
println("Part 2: $(res)")
@assert(res == 16164)

File: 2018/python/day04.py
from enum import Enum, auto
from re import findall


class State(Enum):
    STARTED = auto()
    AWAKE = auto()
    ASLEEP = auto()


class Event:
    def __init__(self, guard, state, beg_sleep):
        self.guard = guard
        self.state = state
        self.beg_sleep = beg_sleep
        self.end_sleep = 0


file = open("data/input-04.txt")
vals = [i.strip() for i in file.readlines()]
vals.sort()

events, guards = [], set()
for idx, line in enumerate(vals):
    minute = int(line[15:17])
    if idx > 0 and events[idx - 1].state == State.ASLEEP:
        events[idx - 1].end_sleep = minute

    case = line[19:24]
    if case == "falls":
        state = State.ASLEEP
        beg_sleep = minute
    elif case == "wakes":
        state = State.AWAKE
        beg_sleep = 0
    elif case == "Guard":
        state = State.STARTED
        [current_guard] = findall("#([0-9]+)", line)
        guards.add(current_guard)
        beg_sleep = 0
    else:
        assert False, "unreachable"

    events.append(Event(current_guard, state, beg_sleep))

guards, sleep, minutes = list(guards), [0] * len(guards), []
for i in range(len(guards)):
    minutes.append([0] * 60)
for event in events:
    [idx] = [idx for idx, guard in enumerate(guards) if guard == event.guard]
    sleep[idx] += event.end_sleep - event.beg_sleep
    for ii in range(event.beg_sleep, event.end_sleep):
        minutes[idx][ii] += 1

max_minute, max_times = [0] * len(guards), [0] * len(guards)
for idx, minute in enumerate(minutes):
    max_minute[idx] = minute.index(max(minute))
    max_times[idx] = max(minute)

guard_id = sleep.index(max(sleep))
res = int(guards[guard_id]) * max_minute[guard_id]
print(f"Part 1: {res}")
assert res == 12169

guard_id = max_times.index(max(max_times))
res = int(guards[guard_id]) * max_minute[guard_id]
print(f"Part 2: {res}")
assert res == 16164
